units/en/unit4/hands-on.mdx

```diff
@@ -0,0 +1,1012 @@
+        img = env.render(mode="rgb_array")
+        images.append(img)
+    imageio.mimsave(out_directory, [np.array(img) for i, img in enumerate(images)], fps=fps)
+```
+
+```python
+def push_to_hub(repo_id, model, hyperparameters, eval_env, video_fps=30, local_repo_path="hub"):
+    """
+    Evaluate, Generate a video and Upload a model to Hugging Face Hub.
+    This method does the complete pipeline:
+    - It evaluates the model
+    - It generates the model card
+    - It generates a replay video of the agent
+    - It pushes everything to the Hub
+
+    :param repo_id: repo_id: id of the model repository from the Hugging Face Hub
+    :param model: the pytorch model we want to save
+    :param hyperparameters: training hyperparameters
+    :param eval_env: evaluation environment
+    :param video_fps: how many frame per seconds to record our video replay
+    :param local_repo_path: where the local repository is
+    """
+
+    _, repo_name = repo_id.split("/")
+    api = HfApi()
+
+    # Step 1: Create the repo
+    repo_url = api.create_repo(
+        repo_id=repo_id,
+        exist_ok=True,
+    )
+
+    # Step 2: Download files
+    repo_local_path = Path(snapshot_download(repo_id=repo_id))
+
+    # Step 3: Save the model
+    torch.save(model, os.path.join(repo_local_path, "model.pt"))
+
+    # Step 4: Save the hyperparameters to JSON
+    with open(Path(repo_local_path) / "hyperparameters.json", "w") as outfile:
+        json.dump(hyperparameters, outfile)
+
+    # Step 5: Evaluate the model and build JSON
+    mean_reward, std_reward = evaluate_agent(
+        eval_env, hyperparameters["max_t"], hyperparameters["n_evaluation_episodes"], model
+    )
+    # Get datetime
+    eval_datetime = datetime.datetime.now()
+    eval_form_datetime = eval_datetime.isoformat()
+
+    evaluate_data = {
+        "env_id": hyperparameters["env_id"],
+        "mean_reward": mean_reward,
+        "n_evaluation_episodes": hyperparameters["n_evaluation_episodes"],
+        "eval_datetime": eval_form_datetime,
+    }
+
+    # Write a JSON file
+    with open(Path(repo_local_path) / "results.json", "w") as outfile:
+        json.dump(evaluate_data, outfile)
+
+    # Step 6: Create the model card
+    # Env id
+    env_name = hyperparameters["env_id"]
+
+    metadata = {}
+    metadata["tags"] = [env_name, "reinforce", "reinforcement-learning", "custom-implementation", "deep-rl-class"]
+
+    # Add metrics
+    eval = metadata_eval_result(
+        model_pretty_name=repo_name,
+        task_pretty_name="reinforcement-learning",
+        task_id="reinforcement-learning",
+        metrics_pretty_name="mean_reward",
+        metrics_id="mean_reward",
+        metrics_value=f"{mean_reward:.2f} +/- {std_reward:.2f}",
+        dataset_pretty_name=env_name,
+        dataset_id=env_name,
+    )
+
+    # Merges both dictionaries
+    metadata = {**metadata, **eval}
+
+    model_card = f"""
```

Comment:
You need to un-indent the following lines, otherwise it will be indented
See https://huggingface.co/osanseviero/reinforce-cartpole-slow  vs https://huggingface.co/osanseviero/reinforce-cartpole-slow2

Reply:
(so un-indenting 757-759)